File: NodeJS/defer/defer.js
/*
 * +===============================================
 * | Author:        Parham Alvani (parham.alvani@gmail.com)
 * |
 * | Creation Date: 02-05-2017
 * |
 * | File Name:     defer.js
 * +===============================================
 */
process.nextTick(() => console.log('-1. tick tack'));

for (let i = 0; i < 10; i++) {
  console.log(`${i}. Hello and How are you bro`);
  process.nextTick(() => console.log(`${i}. tick tack`));
}

File: go/defer/defer.go
/*
 * In The Name Of God
 * ========================================
 * [] File Name : defer.go
 *
 * [] Creation Date : 15-11-2015
 *
 * [] Created By : Parham Alvani (parham.alvani@gmail.com)
 * =======================================
 */
/*
 * Copyright (c) 2015 Parham Alvani.
 */

package main

import (
	"fmt"
	"os"
)

func main() {
	f := createFile("/tmp/defer.txt")
	defer closeFile(f)
	writeFile(f)
}

func createFile(p string) *os.File {
	fmt.Println("creating...")
	f, err := os.Create(p)
	if err != nil {
		panic(err)
	}
	return f
}

func writeFile(f *os.File) {
	fmt.Println("writing...")
	fmt.Fprintln(f, "data")
}

func closeFile(f *os.File) {
	fmt.Println("closing...")
	f.Close()
}
